lib/wallet.py

```diff
@@ -2239,7 +2239,7 @@ def sign_transaction(self, tx, password, *, use_cache=False):
         for k in self.get_keystores():
             try:
                 if k.can_sign(tx):
-                    k.sign_transaction(tx, password, use_cache=use_cache)
+                    k.sign_transaction(tx, password, use_cache=use_cache, ndata=ndata)
```

Comment:
Hmm.  I just checked -- this will break on HW wallets.  

Any new plugin can define a new keystore.  (Usually this is for HW wallets).

Not all possible keystores will have this `ndata` argument -- this will crash on HW wallets.  We have two options here.

1. Restrict the passing-down of  `ndata` only to *known* keystore types.
2. Use Python reflection to detect if the `sign_transaction` method we are calling has the `ndata` kwarg, and if so, pass it down, if not -- give up somehow (either silently fail or warn).

Reply:
I would change this to (for now):

```python
if k.can_sign(tx):
    if ndata:  # used for reusable paycodes
        import inspect
        # test if keystore sign_transaction method knows about the `ndata` kwarg
        if 'ndata' in inspect.signature(k.sign_transaction, follow_wrapped=True).parameters:
           # keystore understands the optional `ndata` kwarg
            k.sign_transaction(tx, password, use_cache=use_cache, ndata=ndata)
        else:
           # keystore does not understand `ndata` (possibly because hw wallet)
            return  # maybe throw exception? warn? idk..
    else:  # regular normal operation
        k.sign_transaction(tx, password, use_cache=use_cache)
```